File: Leetcode/Problems/875. Koko Eating Bananas.cpp
class Solution {
public:
    bool isValid(vector<int>& piles,int mid, int H)
    {
        int h=0;
        for(int pile:piles)
        {
            if(pile%mid==0)
                h+=(pile/mid);
            else
                h = h + (pile/mid) + 1;
            if(h>H)
                return false;
        }
        return true;
    }
    int minEatingSpeed(vector<int>& piles, int H) {
        int beg = 1;
        int end = INT_MIN;
        for(int pile : piles)
        {
            end = max(end,pile);
        }
        
        int result = -1;
        while(beg<=end)
        {
            int mid = beg + (end-beg)/2;
            if(isValid(piles,mid,H))
            {
                result = mid;
                end=mid-1;
            }
            else
            {
                beg = mid+1;
            }
        }
        return result;
    }
};

File: Leetcode/Revision/Binary Search/875. Koko Eating Bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        if(piles.length > h )return -1;
        int maxBan = 0;
        for(int pile:piles)
            maxBan = Math.max(maxBan, pile);
        int start = 1;
        int end = maxBan;
        int k = end;
        while(start<=end){
            int mid = start + (end - start)/2;
            if(isValidK(piles, mid, h)){
                k = mid;
                end = mid-1;
            }
            else
                start = mid+1;
        }
        return k;
    }
    private boolean isValidK(int[] piles, int k, int h){
        int time = 0;
        for(int i=0; i<piles.length; i++){
            if(piles[i]%k ==  0)
                time+=(piles[i]/k);
            else
                time+=((piles[i]/k)+1);
            if(time>h)return false;
        }
        return true;
    }
}
